# Investigation of the FoxRox account on WPN (Winning Poker Network)

Online poker, like many other industries where lots of money is involved, attracts people who will do anything it takes to make money.  Historically, even top poker sites who have dedicated security teams like PokerStars have problems with players cheating via collusion with other players at the table, hacking unsuspecting players, sharing their account with multiple people, or using automated software to make decisions for them (bots).  [In 2015](https://www.parttimepoker.com/pokerstars-plo-bot-ring-is-a-disaster-for-online-poker), a bot ring from Eastern Europe took \\$1.4 million dollars out of the \\$1/2 PLO games on PokerStars.  Game Security is a serious issue for the online poker community because without it, online poker would cease to exist.  It seems shady enough to the average human to deposit \\$50 to an online poker site, and it seems even shadier to do so if they're paranoid about whether or not they're playing against other humans.  Poker players try to do what they can to report offenders, but largely the responsibility of cleaning up the games rests on the poker site itself, as players don't have the ability to ban players themselves.  WPN took an interesting approach to this problem.

WPN is an online poker network, host to several poker sites who merge their players pools together, including: Black Chip Poker, True Poker, and America's Cardroom.  This is an important network in the online poker world because of its inclusion of America's Cardroom, also known as ACR.  ACR is the largest poker site that allows Americans to play (hence its name), therefore making WPN the go-to place to play online poker for Americans, as well being a very nice option for players throughout the rest of the world that have access to WPN.

On April 8th, 2019, Eric Collier, a poker player who streams his gameplay on the popular streaming site [Twitch](https://www.twitch.tv/directory/game/Poker), was heads-up (in the final 2) of a \$15 poker tournament.  Under the his alias "ThaKemist", he ended up [winning](https://www.sharkscope.com/#Find-Tournament//networks/WPN/tournaments/9563589) the tournament for \\$2,388, but not without controversy.  His opponent, FoxRox, exhibited very unhuman-like behavior.  You can read more about the match [here](https://f5poker.com/poker-news/2019/4/11/bot-detected-americas-cardroom-during-twitch-stream/).  The behavior exhibted by the bot isn't our focal point, but how ACR dealt with this issue.  The following images are from a twitter [thread](https://twitter.com/acr_poker/status/1115290003423268865?lang=en), iniated by ThaKemist (@TABLECAPTAINN on twitter) to document his experience with an obvious bot.

---

![Image](https://i.imgur.com/p1oMIrW.jpg)
![Image](https://i.imgur.com/Je2JDPZ.jpg)

---

ACR ultimately [decided](https://www.legaluspokersites.com/news/americas-cardroom-continues-online-poker-bot-crackdown/18460/) to not only make the reimbursement amounts public, but also make some game data from the banned accounts [available to the public](https://www.winningpokernetwork.com/banned-accounts/). Here is a screenshot of some of the top offendors from their website:

---

![Image](https://i.imgur.com/DV1PyN1.jpg)

---

Per their reimbursment policy, the cap reimbursment to the community per offender is \\$25,000, except for our friend FoxRox.  Over \\$175k was given back to players affected by FoxRox on WPN, with 1.3% of players on the site haven been cheated by this player (this is probably a rather large number, in perhaps the tens of thousands of players).  Under the Downloads column for FoxRox, we have a couple of Excel files containing game data from his tournament play.  Let's open them up, shall we?

In [51]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import datetime as dt

In [52]:
mtt_one = pd.read_csv('wpn-foxrox-file1.csv')
print("Number of rows: "+str(len(mtt_one)))
mtt_one.head(5)

Number of rows: 553101


TournamentName  TournamentId       StartDate          Nickname  \
0  PHAT BIG10 - $15,000 GTD       8738624  6/15/2018 0:00    GrinderTiltOff   
1  PHAT BIG10 - $15,000 GTD       8738624  6/15/2018 0:00        SimSalabim   
2  PHAT BIG10 - $15,000 GTD       8738624  6/15/2018 0:00          lepewpew   
3  PHAT BIG10 - $15,000 GTD       8738624  6/15/2018 0:00  HipllteMeRunoGod   
4  PHAT BIG10 - $15,000 GTD       8738624  6/15/2018 0:00          MN_KVANG   

   FinishPlace    Award  
0            1  3587.50  
1            2  2231.25  
2            3  1575.00  
3            4  1172.50  
4            5   910.00

In [53]:
mtt_one.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 553101 entries, 0 to 553100
Data columns (total 6 columns):
TournamentName    553101 non-null object
TournamentId      553101 non-null int64
StartDate         553101 non-null object
Nickname          553101 non-null object
FinishPlace       553101 non-null int64
Award             553101 non-null float64
dtypes: float64(1), int64(2), object(3)
memory usage: 25.3+ MB


Wow, 553,000+ rows, and it's only the first file.  It seems to carry not only the results of FoxRox, but the results of his opponents in all the tournaments he played in as well.  This is merely an assumption at the moment, let's try compiling and sorting the data to confirm this.

In [54]:
mtt_two = pd.read_csv('wpn-foxrox-file2.csv')
main_pd = pd.concat([mtt_one, mtt_two], ignore_index=True)

main_pd.tail(5)

TournamentName  TournamentId      StartDate      Nickname  \
1062892   $500 GTD R/A       9564347  4/8/2019 5:00    robmcguire   
1062893   $500 GTD R/A       9564347  4/8/2019 5:00   joydivizion   
1062894   $500 GTD R/A       9564347  4/8/2019 5:00  vegasbound10   
1062895   $500 GTD R/A       9564347  4/8/2019 5:00     qu33n b33   
1062896   $500 GTD R/A       9564347  4/8/2019 5:00      jgallo23   

         FinishPlace  Award  
1062892           70    0.0  
1062893           71    0.0  
1062894           72    0.0  
1062895           73    0.0  
1062896           74    0.0

In [55]:
main_pd.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1062897 entries, 0 to 1062896
Data columns (total 6 columns):
TournamentName    1062897 non-null object
TournamentId      1062897 non-null int64
StartDate         1062897 non-null object
Nickname          1062897 non-null object
FinishPlace       1062897 non-null int64
Award             1062897 non-null float64
dtypes: float64(1), int64(2), object(3)
memory usage: 48.7+ MB


Now that we have combined the dataframes to over 1,000,000 rows, let's think about which parts of the data will be useful:
* **TournamentName** and **TournamentID** both seem useful in comparing results from FoxRox to other players, as we can use the df.groupby() or df.pivot_table() methods for this.  TournamentName is the name of the tournament (many of these are daily/weekly), and TournamentID is the individual ID of a specific tournament.  One TournamentName will have a new TournamentID every time it runs.
* **StartDate** will be useful in visualizing FoxRox and other player's winnings over time.  It is currently in string format, and we will conver this into datetime format before filtering further.
* **Nickname** is obviously a critical column, and we can filter results with this column the same way we can with Tournament Name/ID.
* **FinishPlace** and **Award** show the place in which the player (nickname) finished and how much money was won, respectively.  FinishPlace is in integer format, and Award is in float format.

Let's convert the StartDate column to datetime format.

In [56]:
# make new series with just the converted datetimes
just_dates = pd.to_datetime(main_pd['StartDate'])
just_dates.head(5)

0   2018-06-15
1   2018-06-15
2   2018-06-15
3   2018-06-15
4   2018-06-15
Name: StartDate, dtype: datetime64[ns]

In [57]:
# change StartDate to the converted datetimes
main_pd['StartDate'] = just_dates
print(type(main_pd.iloc[0,2]))
main_pd.head(5)

<class 'pandas._libs.tslibs.timestamps.Timestamp'>


TournamentName  TournamentId  StartDate          Nickname  \
0  PHAT BIG10 - $15,000 GTD       8738624 2018-06-15    GrinderTiltOff   
1  PHAT BIG10 - $15,000 GTD       8738624 2018-06-15        SimSalabim   
2  PHAT BIG10 - $15,000 GTD       8738624 2018-06-15          lepewpew   
3  PHAT BIG10 - $15,000 GTD       8738624 2018-06-15  HipllteMeRunoGod   
4  PHAT BIG10 - $15,000 GTD       8738624 2018-06-15          MN_KVANG   

   FinishPlace    Award  
0            1  3587.50  
1            2  2231.25  
2            3  1575.00  
3            4  1172.50  
4            5   910.00

In [58]:
# making a dataframe that shows only entries for FoxRox
fr_pd = main_pd[main_pd["Nickname"] == "FoxRox"]
fr_pd.drop_duplicates()
fr_pd.head(5)

TournamentName  TournamentId           StartDate  \
87           PHAT BIG10 - $15,000 GTD       8738624 2018-06-15 00:00:00   
262                        $2,000 GTD       8738673 2018-06-15 00:10:00   
599   TURBO10 - $2,000 GTD Superstack       8738726 2018-06-15 00:30:00   
775          BIG10 - $2,000 GTD 6 Max       8738802 2018-06-15 01:00:00   
1010                       $5,000 GTD       8738867 2018-06-15 01:15:00   

     Nickname  FinishPlace  Award  
87     FoxRox           88   0.00  
262    FoxRox           29  13.89  
599    FoxRox           43   0.00  
775    FoxRox          100   0.00  
1010   FoxRox          109   0.00

In [59]:
fr_pd.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 4587 entries, 87 to 1062888
Data columns (total 6 columns):
TournamentName    4587 non-null object
TournamentId      4587 non-null int64
StartDate         4587 non-null datetime64[ns]
Nickname          4587 non-null object
FinishPlace       4587 non-null int64
Award             4587 non-null float64
dtypes: datetime64[ns](1), float64(1), int64(2), object(2)
memory usage: 250.9+ KB


There are 4,587 tournament result entries here for FoxRox. This seems to indicate that we have a near complete or complete set of results for this player, especially when you match it up with their [sharkscope results](https://www.sharkscope.com/#Player-Statistics//networks/WPN/players/FoxRox) which showed 4,534 results, a fairly close number to the one we have.

In [63]:
# grouping data, starting with TournamentName

fr_tournament_pivot = fr_pd.pivot_table(values=["FinishPlace", "Award"], index='TournamentName')
fr_tournament_pivot.head(5)

Award  FinishPlace
TournamentName                                             
$1,000 GTD                           24.451000   132.800000
$1,000 GTD - Hyper Turbo MEGA STACK   2.566667    90.733333
$1,000 GTD 6-Max                     48.375000    21.250000
$1,000 GTD 8 max                      6.490000    29.000000
$1,000 GTD 8-Max                     39.304390    78.951220

Looking above,fix names

In [64]:
# average finish
fr_avg_finish = fr_pd['FinishPlace'].mean()
print(fr_avg_finish)

# average award and highest
fr

# longest losing streak

# total won, biggest profit day, worst losing day

# profit by day and month

# avg profit per tournament, total profit per tournament

# 

110.72204054937868


NameError: name 'fr' is not defined